### DISCUSSION TITLE
Doubts about GraphQL Federation 2 new directives

### DISCUSSION THREAD

**Discussion by valeriob23** [Netflix/dgs-framework#1136] [Q&A]
Hello everyone,
I'm trying to federate my service to an Apollo Supergraph using Federation v2 that, from what I understand, is [fully supported](https://www.apollographql.com/docs/federation/other-servers/#java--kotlin) by DGS.
Specifically, I need to reference an entity without contributing fields, therefore using the `resolvable` argument with the `@key` directive as suggested [here](https://www.apollographql.com/docs/federation/entities/#referencing-an-entity-without-contributing-fields). You can find how my `schema.graphqls` file should look (simplified version) below, where User is actually a type defined in another subgraph/service:

```graphql
type Query {
    survey: Survey
}

type User @key(fields: "id", resolvable: false) {
    id: ID!
}

type Survey {
    title: String
    author: User
}
```

Unfortunately, if I try to generate code and then run the service, it complains this way:
```
There are problems with the GraphQL Schema:
	 * 'User' [@26:1] use an unknown argument 'resolvable' on directive 'key'
```

On the other hand, trying to force the import of new Federation 2 directives as follows
```graphql
extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@shareable"])
```
my DGS service doesn't complain anymore; however, when I try to generate the supergraph, I obtain the following error:
```
Invalid definition for directive "@key": argument "fields" should have type "_FieldSet!" but found type "federation__FieldSet!"
```

Just another information: I didn't define any directives as indicated [here](https://www.apollographql.com/docs/federation/federation-spec#federation-schema-specification) as, correct me if I'm wrong, DGS already follow the latest specification.

Could you please help me in understading where I'm making a mistake and achieve what I've explained?
Thanks!!

### Context ###

- Language: Kotlin (1.7)
- DGS version: 5.0.5

**Comment by srinivasankavitha** [comment]
Hi! Thanks for trying this out and reporting. While we've added federation
v2.0 in the library, but have not yet fully vetted all our related tooling
to support this. We plan to do so in the coming weeks, but until we do
expect that not everything will work as expected.

**Reply by valeriob23** [reply]
hey, thanks for replying. Do you think I should open an issue, or it is something already addressed?

**Reply by srinivasankavitha** [reply]
Feel free to open issues with any specific problems you run into. That will help us as well while we work on this effort.

**Reply by lbusch25** [reply]
@srinivasankavitha any updates on the progress of this effort? 

My team just ran into the same issue as @valeriob23 while trying to upgrade to Federation 2.0 using the Apollo Router.

**Reply by srinivasankavitha** [reply]
We haven't been able to prioritize looking into federation v2.0 yet unfortunately. We are not updating internally so this effort is deferred. It is backwards compatible, just the new features are not tried and tested. I expect we will be able to look into this in a couple months when we actively start using it within the company.

**Reply by lbusch25** [reply]
Thanks for the rapid reply @srinivasankavitha. That is understandable.

One additional piece of information that I can provide is that the workaround listed below doesn't appear to work with the Apollo Router (although we have admitted done minimal exploration on that front). With the Apollo Gateway we were able to get DGS working with the `@key` directive on federation 2.0, although with the additional configuration mentioned below. 

For now my team can just move to the router and stay with federation `1.x`, but I think the team from apollo should be notified if possible so that their docs do not state DGS as 2.0 compatible as of yet (At least not fully).

**Reply by srinivasankavitha** [reply]
That's great to know and thanks for the details. I will follow up with the folks from Apollo to clarify that and work with them.

**Reply by setchy** [reply]
@lbusch25 in addition, may I also suggest raising an issue in https://github.com/apollographql/apollo-federation-subgraph-compatibility if you believe there is differing computability between apollo router and Apollo gateway

**Comment by Rfank2021** [comment]
I'm trying to use `@shareable`, but seems not work.
```'schema' [@-1:-1] tried to use an undeclared directive 'link'```

**Reply by Rfank2021** [reply]
Need upgrade dgs libraries, also includes all directives in the import.

**Comment by basaa** [comment]
I imported `@shareable` from `extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["FieldSet", "@key", "@shareable", "@external", "@provides"])` but it seems not to work too.

I get the following errors:
`UNKNOWN: [accounts] Invalid definition for directive "@key": argument "fields" should have type "_FieldSet!" but found type "FieldSet!"`
    `UNKNOWN: [accounts] Invalid definition for directive "@provides": argument "fields" should have type "_FieldSet!" but found type "FieldSet!"`

Packages:
`"@apollo/rover": "^0.7.0"`
`"@apollo/subgraph": "^2.0.5"`
`"@apollo/composition": "^2.0.5"`
`"@apollo/federation": "^0.36.1"`
`"@apollo/federation-internals": "^2.0.5"`
`"@apollo/gateway": "^2.0.5"`

**Reply by valeriob23** [reply]
You could try to extend the schema the following way (I omit some imports to make it easier):
```graphql
extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: [{ name: "FieldSet", as: "_FieldSet"}, "@key"])
```
Just tried and it seems the supergraph generation doesn't complain anymore. I didn't try if everything work as expected, though. Could you check also on your side?

**Reply by basaa** [reply]
It worked! 🚀

However I have a weird issue,  when I use `@shareable` on `Account` (object) and compose a Supergraph, all `Account`'s fields are required to be resolve on all subgraphs (which is quite annoying to maintain), but if I use `IntrospectAndCompose` everything is works fine, maybe I can solve it with a better resolver strategy.

Thanks for the tip @valeriob23

**Reply by Rfank2021** [reply]
What do you mean? By using `IntrospectAndCompose` you don't need to `resolve on all subgraphs`?

**Reply by basaa** [reply]
> What do you mean? By using `IntrospectAndCompose` you don't need to `resolve on all subgraphs`?

No you don't need to, only the fields that are present in your `@key`

**Reply by Rfank2021** [reply]
Yeah, that's how federation works.

**Reply by basaa** [reply]
#### USERS SUBGRAPH
```
type User @key(fields: "id") @shareable {
  id: ID!
  name: String
}
```

#### PRODUCTS SUBGRAPH
```
type Product @key(fields: "upc") @shareable {
  upc: UPC!
  name: String
  description: String
}

type User @key(fields: "id") @shareable {
  id: ID!
  email: String
  name: String
}
```

This configuration works with `IntrospectAndCompose` but doesn't with `rover supergraph compose`
```
from subgraph "users":
 - cannot find field "User.email".
 - cannot move to subgraph "products", which has field "User.email", because type "User" has no @key defined in subgraph "products"
```

maybe I'm missing something...
